setup.ps1

```diff
@@ -0,0 +1,229 @@
+  )
+  
+  Write-Log "$prompt`n" -Level 'PROMPT'
+  for ($i = 0; $i -lt $options.Length; $i++) {
+    Write-Log "$([char](65 + $i)). $($options[$i])" -Level 'PROMPT'
+  }
+  
+  Write-Log "`nSelect one or more options by typing the corresponding letters, separated by commas." -Level 'PROMPT'
+  $inputPath = Read-Host 
+  if ([string]::IsNullOrEmpty($inputPath)) {
+    $inputPath = $defaultChoice
+  }
+  
+  # Ensure $inputPath is treated as a string and split it by commas
+  $inputPath = [string]$inputPath
+  $selections = $inputPath.Split(',') | ForEach-Object {
+    $_.Trim().ToUpper()
+  }
+  
+  # Convert each selection from letter to index and validate
+  $indices = @()
+  foreach ($selection in $selections) {
+    if ($selection -match '^[A-Z]$') {
+      $index = [int][char]$selection - [int][char]'A'
+      if ($index -ge 0 -and $index -lt $options.Length) {
+        $indices += $index
+      }
+      else {
+        Write-Log "Invalid input: $selection. Please enter letters within the valid range of options." -Level 'ERROR'
+      }
+    }
+    else {
+      Write-Log "Invalid input: $selection. Please enter letters between A and Z." -Level 'ERROR'
+    }
+  }
+  
+  return $indices
+}
+
+function Exit-Script {
+  param (
+    [int]$exitCode,
+    [bool]$isSubShell = $true,
+    [bool]$waitForKeypress = $true
+  )
+
+  if ($OldVirtualPath) {
+    $env:PATH = $OldVirtualPath
+  }
+
+  if ($exitCode -ne 0 -and $isSubShell) {
+    Write-Log "Script failed. Would you like to open the log file? (Y/N)" -Level 'PROMPT'
+    $openLog = Read-Host
+    if ($openLog -eq 'Y' -or $openLog -eq 'y') {
+      Start-Process notepad.exe -ArgumentList $LogFilePath
+    }
+  }
+
+  if ($waitForKeypress) {
+    Write-Log "Press any key to exit..."
+    $host.UI.RawUI.ReadKey("NoEcho,IncludeKeyDown") | Out-Null
+  }
+
+  return $exitCode
+}
+
+# Function to handle installation
+function Install {
+  param ([string]$InputPath)
+  
+  if (-not $InputPath) {
+    Write-Log "No input provided for installation." -Level 'ERROR'
+    Exit-Script -exitCode 1
+  }
+
+  Write-Log "Installing $InputPath..."
+  $installCmd = if (Test-Path $InputPath) { $VenvPip + @('install', '-r', $InputPath) } else { $VenvPip + @('install', $InputPath) }
+  $output = & $installCmd[0] $installCmd[1..$installCmd.Length] 2>&1
+  $output | Out-File $LogFilePath -Append
+  if ($LASTEXITCODE -ne 0) {
+    Write-Log "Conflict detected. Exiting now..." -Level 'ERROR'
+    Exit-Script -exitCode 1
+  }
+}
+
+# Function to get the installed Python version tag for wheel compatibility
+function Get-PythonVersionTag {
+  $pythonVersion = & python -c "import sys; print(f'cp{sys.version_info.major}{sys.version_info.minor}')"
+  $architecture = & python -c "import platform; print('win_amd64' if platform.machine().endswith('64') else 'win32')"
+  return "$pythonVersion-$architecture"
+}
+
+# Exit in test environment
+if ($MyInvocation.InvocationName -ne $MyInvocation.MyCommand.Name) {
```

Comment:
This seems flaky (at best) ... and fails when called with `.\setup.ps1`.

it also seems to fail in when called with `powershell -File setup.ps1` - so i'm really not sure how this is intended to be called ... ?

Reply:
It is to prevent that the main routine gets executed when the functions need to be called from the unit test file.

Reply:
I always call the setup file another way, so didn't notice that behaviour you mentioned.

Reply:
well the intend is clear i think - but it doesn't really work for non-test cases ... at least not for me  :laughing: 

Now i'm curious - how are you calling this?
i found no way to call it to actually have it pass this check ... - not using `.\setup.ps1`, `% setup.ps1` - and not with the `-File` method mentioned above - ... so neither me, nor a quick google search were able to call this in a way to make it work.

Reply:
Somehow it stopped working for me too (maybe because of changing file path). I guess I had a "lucky" situation in the moment where it worked. 😂 I decided to export the functions in a psm file to be able to call the functions separately for unit testing.

Reply:
I am also not a fan of it. At that point, I had no other idea. I'll try out your suggestion with mocking the main function. Thanks.